components/raftstore/src/store/fsm/apply.rs

```diff
@@ -3589,6 +3687,11 @@ where
     #[cfg(any(test, feature = "testexport"))]
     #[allow(clippy::type_complexity)]
     Validate(u64, Box<dyn FnOnce(*const u8) + Send>),
+    ApplyGap {
```

Comment:
What is `ApplyGap`?

Reply:
Eg:

T0 Loop 1: The witness handles `ready.committed_entries()`, gets `entries [6, 20]`, and the index of `switch to non-witness` is `9`. After the `witness` peer applies `9`, it will skip apply `[10, 20]`.

T1 The leader will send a snapshot, its index may between `[10, 20]`, eg `12`.

T2 Loop 2: The applied snapshot `non-witness` handle next  `ready.committed_entries()`, gets `entries [21, 40]`.

So there is a log gap `[13, 20]` without `apply`. The `ApplyGap` is for `apply` this `gap`.

Reply:
I think it will panic at T1 as shown in:

https://github.com/tikv/raft-rs/blob/5ce52b480065ff31ecef16b9b77c7c3b7c57c8c7/src/raw_node.rs#L520-L529

Reply:
I see,  so I may need to make non-witness `request_index` to T0's `the lastest committed_index`,  and forbid persistence of new raft log entries (maybe they were synchronized before the leader received the snapshot request), so even leader updated and sync `new committed index` to follower, the follower's raft `persisted` still equal to  `the prev_lastest committed_index` to make https://github.com/tikv/raft-rs/blob/5ce52b480065ff31ecef16b9b77c7c3b7c57c8c7/src/raw_node.rs#L520-L529 it pass?

Reply:
> I see, so I may need to make non-witness `request_index` to `the lastest committed_index`, and forbid persistence of new raft log entries (maybe they were synchronized before the leader received the snapshot request), so even leader updated and sync `new committed index` to non-witness, the non-witness's raft `persisted` still equal to `the prev_lastest committed_index` to make https://github.com/tikv/raft-rs/blob/5ce52b480065ff31ecef16b9b77c7c3b7c57c8c7/src/raw_node.rs#L520-L529 pass?

I'll think more about it.

Reply:
Maybe `last_applying_idx` should be decreased if a snapshot is received and accepted.

Reply:
Make non-witness `request_index` to its committed_index, and reject raft append msg when in `wait_data`. And on leader side, because gen-snapshot tasks go through apply worker, the snapshot's index must be equal to or larger than non-witness's committed_index. Then `ApplyGap` is not needed and won't encounter the assertion.

Reply:
> because gen-snapshot tasks go through apply worker, the snapshot's index must be equal to or larger than non-witness's committed_index

It's not true when paging is enabled for committed entries.

Reply:
> > because gen-snapshot tasks go through apply worker, the snapshot's index must be equal to or larger than non-witness's committed_index
> 
> It's not true when paging is enabled for committed entries.

When gen snapshot, it will check snapshot's idx and `request_index`: https://github.com/tikv/tikv/blob/master/components/raftstore/src/store/peer_storage.rs#L407
So if we set `request_index` to non-witness `latest comit index` (after `wait_data` set to `true`, reject `MsgAppend`), it should be ok?

Reply:
@BusyJay What's paging?

Reply:
Can we just remove the assertion?

Reply:
I think there are two problems:

1. Requesting latest committed logs are not enough because leader may just keep sending committed entries to witness and increase its `commit_since_idx`. Reject `MsgAppend` will make a raft group with one follower isolated unavailable immediately. The assert can help find potential flaws in `Storage` implementations. I would suggest to make the assert more strict that allows explicit requested snapshots. Note `commit_since_idx` is just one of the case, I think raft_log.persisted will also be affected and maintained carefully.
2. Skipping applying logs may make the follower unable to compaign when there are pending conf changes. So if the wait data non-witness has the most up to date logs when leader is isolated, the group will also be unavailable.

The simplest solution is to request the snapshot at last_index and then reject any logs afterwards. The term should be also attached to the snapshot requests, leader generates snapshot if `local_term > request_term || local_index >= request_index`. This solution will not solve the unavailability corner cases described above, but is safest.

In opinion, switch witness is no different from remove the node and then add a new node except it can still campaign if leader is isolated during the time. So it has a little higher availability then remove + add.